shell/platform/darwin/macos/framework/Source/FlutterEmbedderKeyResponder.mm

```diff
@@ -0,0 +1,715 @@
+      uint64_t secondCode = (uint64_t)[keyLabel characterAtIndex:1];
+      codeUnit = (codeUnit << 16) | secondCode;
+    }
+    if (codeUnit < 256) {
+      return toLower(codeUnit);
+    }
+    return KeyOfPlane(codeUnit, kUnicodePlane);
+  }
+
+  // Control keys like "backspace" and movement keys like arrow keys don't have
+  // a printable representation, but are present on the physical keyboard. Since
+  // there is no logical keycode map for macOS (macOS uses the keycode to
+  // reference physical keys), a LogicalKeyboardKey is created with the physical
+  // key's HID usage and debugName. This avoids duplicating the physical key
+  // map.
+  if (physicalKey != 0) {
+    return KeyOfPlane(physicalKey, kHidPlane);
+  }
+
+  // This is a non-printable key that is unrecognized, so a new code is minted
+  // with the autogenerated bit set.
+  return KeyOfPlane(event.keyCode, kMacosPlane | kAutogeneratedMask);
+}
+
+/**
+ * Converts NSEvent.timestamp to the timestamp for Flutter.
+ */
+static double GetFlutterTimestampFrom(NSTimeInterval timestamp) {
+  // Timestamp in microseconds. The event.timestamp is in seconds with sub-ms precision.
+  return timestamp * 1000000.0;
+}
+
+/**
+ * Compute |modifierFlagOfInterestMask| out of |keyCodeToModifierFlag|.
+ *
+ * This is equal to the bitwise-or of all values of |keyCodeToModifierFlag| as
+ * well as NSEventModifierFlagCapsLock.
+ */
+static NSUInteger computeModifierFlagOfInterestMask() {
+  __block NSUInteger modifierFlagOfInterestMask = NSEventModifierFlagCapsLock;
+  [keyCodeToModifierFlag
+      enumerateKeysAndObjectsUsingBlock:^(NSNumber* keyCode, NSNumber* flag, BOOL* stop) {
+        modifierFlagOfInterestMask = modifierFlagOfInterestMask | [flag unsignedLongValue];
+      }];
+  return modifierFlagOfInterestMask;
+}
+
+/**
+ * The C-function sent to the embedder's |SendKeyEvent|, wrapping
+ * |FlutterEmbedderKeyResponder.handleResponse|.
+ *
+ * For the reason of this wrap, see |FlutterKeyPendingResponse|.
+ */
+void HandleResponse(bool handled, void* user_data);
+
+/**
+ * Converts NSEvent.characters to a C-string for FlutterKeyEvent.
+ */
+const char* getEventString(NSString* characters) {
+  if ([characters length] == 0) {
+    return nullptr;
+  }
+  unichar utf16Code = [characters characterAtIndex:0];
+  if (utf16Code >= 0xf700 && utf16Code <= 0xf7ff) {
+    // Some function keys are assigned characters with codepoints from the
+    // private use area (see
+    // https://developer.apple.com/documentation/appkit/1535851-function-key_unicodes?language=objc).
+    // These characters are filtered out since they're unprintable.
+    //
+    // Although the documentation claims to reserve 0xF700-0xF8FF, only up to
+    // 0xF747 is actually used. Here we choose to filter out 0xF700-0xF7FF
+    // section. The reason for keeping the 0xF800-0xF8FF section is because
+    // 0xF8FF is used for the "Apple logo" character (Option-Shift-K on US
+    // keyboard.)
```

Comment:
Can you document here why you're including the 0xF800-0xF8FE region? e.g. "The 0xF800-0xF8FE region is included in anticipation of any new private values that are printable."

Do you have any documentation that this is how they split it up, or is that just a feeling? If it's not just a feeling, including a link to that information would be good.

I'm still a little leery of making the assumption that the region will include printable characters, but the consequences aren't huge, so I guess it's OK.

Reply:
I understand. It is indeed only a feeling, and a result out of experimentation. The only documentation I can find is the one I included that reserves 0xF700-0xF8FF. Since we're trusting the official documentation, the doc only declared a limited number of values (https://developer.apple.com/documentation/appkit/1535851-function-key_unicodes?language=objc), all of which are within 0xF700-0xF747. I'll add a little more to the doc. We can always come back and adjust this range anyway.

Reply:
How does this look?
```
    // Some function keys are assigned characters with codepoints from the
    // private use area. These characters are filtered out since they're
    // unprintable.
    //
    // The official documentation reserves 0xF700-0xF8FF as private use area
    // (https://developer.apple.com/documentation/appkit/1535851-function-key_unicodes?language=objc).
    // But macOS seems to only use a reduced range of it. The official doc
    // defines a few constants, all of which are within 0xF700-0xF747.
    // (https://developer.apple.com/documentation/appkit/1535851-function-key_unicodes?language=objc).
    // This mostly aligns with the experimentation result, except for 0xF8FF,
    // which is used for the "Apple logo" character (Option-Shift-K on a US
    // keyboard.)
    //
    // We hereby assume that non-printable function keys are defined from
    // 0xF700 upwards, and printable private keys are defined from 0xF8FF
    // downwards. Since we want to keep the printable private keys, therefore
    // we only filter out 0xF700-0xF7FF.
```